torch/_inductor/memory.py

```diff
@@ -0,0 +1,543 @@
+from __future__ import annotations
+
+import dataclasses
+import heapq
+import logging
+from typing import Callable, Dict, List, Set, Tuple, TYPE_CHECKING, Union
+
+from torch.utils._ordered_set import OrderedSet
+
+from .ir import MultiOutputLayout
+from .utils import get_dtype_size
+from .virtualized import V
+
+
+if TYPE_CHECKING:
+    from .dependencies import Dep
+    from .scheduler import BaseSchedulerNode, SchedulerBuffer
+
+
+torch_log = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class MemoryPlanningInfoForBuffer:
+    size_alloc: int = 0
+    size_free: int = 0
+    outdegree: int = 0
+    succ_nodes: OrderedSet[BaseSchedulerNode] = dataclasses.field(
+        default_factory=OrderedSet
+    )
+
+
+@dataclasses.dataclass
+class MemoryPlanningInfoForNode:
+    pred_buffers: List[Union[SchedulerBuffer, FreeableInputBuffer]] = dataclasses.field(
+        default_factory=list
+    )
+    pred_nodes: List[BaseSchedulerNode] = dataclasses.field(default_factory=list)
+    succ_nodes: List[BaseSchedulerNode] = dataclasses.field(default_factory=list)
+    indegree: int = 0
+    index: int = 0
+    memory_to_free: int = 0
```

Comment:
is `memory_to_free` only used in LPMF? wonder if we can make it internal to that algorithm function

Reply:
I was doing so before. So instead of declaring `memory_to_free` as an attribute in this dataclass, I dynamically set it inside the `_lpmf()` function. However, mypy linter is throwing errors. I am not sure how to not pre-declare these and also make linter happy :( We can discuss more about this.

Reply:
sounds great - maybe a comment to call out it's lpmf only is sufficient